internal/resource/route.go

```diff
@@ -102,17 +116,58 @@ func (r Route) ProcessDefaults(ctx context.Context) error {
 		return fmt.Errorf("invalid nil resource")
 	}
 	if len(r.Route.Protocols) == 0 {
-		err := mergo.Merge(r.Route, defaultRoute,
-			mergo.WithTransformers(wrappersPBTransformer{}))
-		if err != nil {
-			return err
+		if r.Route.Expression == "" {
+			err := mergo.Merge(r.Route, defaultRoute,
+				mergo.WithTransformers(wrappersPBTransformer{}))
+			if err != nil {
+				return err
+			}
+		} else {
+			err := mergo.Merge(r.Route, defaultExpressionRoute,
+				mergo.WithTransformers(wrappersPBTransformer{}))
+			if err != nil {
+				return err
+			}
 		}
 	}
 	defaultID(&r.Route.Id)
 	return nil
 }
 
+func buildSchema() *goatcrouter.Schema {
+	schema := goatcrouter.NewSchema()
+
+	for _, fieldname := range []string{
+		"net.protocol", "tls.sni",
+		"http.method", "http.host",
+		"http.path", "http.raw_path",
+		"http.headers.*",
+	} {
+		schema.AddField(fieldname, goatcrouter.String)
+	}
+
+	for _, fieldname := range []string{"net.port"} {
+		schema.AddField(fieldname, goatcrouter.Int)
+	}
+
+	return schema
+}
+
+var cachedSchema = buildSchema()
+
 func init() {
+	jsonschema.Formats["route-expression"] = func(v interface{}) bool {
+		expression, ok := v.(string)
+		if !ok {
+			return false
+		}
+
+		router := goatcrouter.NewRouter(cachedSchema)
+		defer router.Free()
+		err := router.AddMatcher(0, uuid.Must(uuid.NewV4()), expression)
```

Comment:
Why the random UUID?

Reply:
the `AddMatcher()` method requires an UUID to handle expressions.  the Lua code uses the route entity id, but in this JsonSchema callback we don't have access to other fields.

We don't actually use the id because we just want to verify that the expression gets compiled without error and drop the whole router object immediately.  since each router only ever gets a single expression, we could use the same predefined id every time.  either by moving the uuid creation to the init phase, or by using a literal constant.

but if performance results a real concern, we should get the internal Rust `parse` and `ast.validate` functions exposed and thus avoid the creation of a router object.

Reply:
Got it, makes sense, just seemed a bit strange while reading the code. Appreciate the explanation. I'm assuming there's not much of a performance improvement using atc-router's own internal `parse()` & `ast.validate()` functions.